<a href="https://colab.research.google.com/github/lrsilsbee/Machine-Learning/blob/main/Lauren_Silsbee_Deep_Learning_Challenge.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
pip install tensorflow-gpu==2.0.0-rc1

     |████████████████████████████████| 380.5MB 38kB/s 
     |████████████████████████████████| 51kB 7.9MB/s 
     |████████████████████████████████| 4.3MB 51.6MB/s 
     |████████████████████████████████| 501kB 60.9MB/s 


In [ ]:
#imports
import warnings
warnings.filterwarnings("ignore")

from tensorflow.keras.datasets import fashion_mnist
from tensorflow.keras.utils import to_categorical
from tensorflow.keras.models import Sequential 
from tensorflow.keras.layers import Dense
from tensorflow.keras import optimizers

1. Preprocess your data so that you can feed it into ANN models.
2. Split your data into training and test sets.
3. Compare your models' training scores and interpret your results.
4. Evaluate your models' performances on your test set.
5. Compare the results of your models.

The Fashion MNIST dataset is very similar to the MNIST dataset, as it has 60,000 training observations, 10,000 testing observations, 10 classes, and then 28x28 images. So, the preprocessing can be similar to the preprocessing steps used for the MNIST dataset.

In [ ]:
(X_train, y_train), (X_test, y_test) = fashion_mnist.load_data()
#Fashion MNIST has 28x28 greyscale images, so 28x28 = 784
input_dim = 784  
#10 classes:
output_dim = nb_classes = 10
#set 20 epochs
nb_epoch = 20

#split into test and training sets
X_train = X_train.reshape(60000, input_dim)
X_test = X_test.reshape(10000, input_dim)
#convert to float type
X_train = X_train.astype('float32')
X_test = X_test.astype('float32')
#normalize vectors
X_train /= 255
X_test /= 255

4423680/4422102 [==============================] - 0s 0us/step


In [ ]:
#convert Y_train, Y_test to categorical type
Y_train = to_categorical(y_train, nb_classes)
Y_test = to_categorical(y_test, nb_classes)


Try different ANN models and train them on your training set.
You can play with:



> 3.1. Number of layers.

> 3.2. Activation functions of the layers.

>3.3. Number of neurons in the layers.

>3.4. Different batch sizes during training.

Let's start with a model from the checkpoint that worked well with the MNIST, just as a starting point: 3 dense layers with 128, 64, and 10 neurons, with batch size set to 128, and then tweak the number of layers.

In [ ]:
#build the model
model = Sequential()
# add 3 dense layers with 128, 64, 10 neurons
model.add(Dense(128, input_shape=(784,), activation="relu"))
model.add(Dense(64, activation="relu"))
model.add(Dense(10, activation="softmax"))
#compile the model
model.compile(optimizer='sgd', loss='categorical_crossentropy',
              metrics=['accuracy'])
#set 128 as mini batch size
model.fit(X_train, Y_train, batch_size=128, epochs=20, verbose=False)
#evaluate the model
score = model.evaluate(X_test, Y_test, verbose=0)
print('Test score:', score[0])
print('Test accuracy:', score[1])

Test score: 0.44902896881103516
Test accuracy: 0.8364999890327454


The above model had a decent accuracy score and does not seem to suffer from overfitting. I'll add another layer and see how they compare.

In [ ]:
#build the model
model = Sequential()
# add 4 dense layers with 128, 64, 58, 10 neurons
model.add(Dense(128, input_shape=(784,), activation="relu"))
model.add(Dense(64, activation="relu"))
model.add(Dense(58, activation="relu"))
model.add(Dense(10, activation="softmax"))
#compile the model
model.compile(optimizer='sgd', loss='categorical_crossentropy',
              metrics=['accuracy'])
#set 128 as mini batch size
model.fit(X_train, Y_train, batch_size=128, epochs=20, verbose=False)
#evaluate the model
score = model.evaluate(X_test, Y_test, verbose=0)
print('Test score:', score[0])
print('Test accuracy:', score[1])

Test score: 0.42645224928855896
Test accuracy: 0.8454999923706055


The results for models 1 and 2 are incredibly similar for both test and training set accuracy scores. I'll add a fifth layer and see how that stacks up.

In [ ]:
#build the model
model = Sequential()
# add 5 dense layers with 128, 100, 64, 58, 10 neurons
model.add(Dense(128, input_shape=(784,), activation="relu"))
model.add(Dense(100, activation="relu"))
model.add(Dense(64, activation="relu"))
model.add(Dense(58, activation="relu"))
model.add(Dense(10, activation="softmax"))
#compile the model
model.compile(optimizer='sgd', loss='categorical_crossentropy',
              metrics=['accuracy'])
#set 128 as mini batch size
model.fit(X_train, Y_train, batch_size=128, epochs=20, verbose=False)
#evaluate the model
score = model.evaluate(X_test, Y_test, verbose=0)
print('Test score:', score[0])
print('Test accuracy:', score[1])

Test score: 0.38783755898475647
Test accuracy: 0.8615000247955322


Adding the fifth layer very slightly increased the test score's accuracy but the training score lowered. I'll stick with 3 layers since the accuracy scores are all similar and this way there are fewer parameters to estimate.

Now, I want to see how varying the activation function changes the results. I'll compare relu, tanh, and sigmoid activation functions.

Using ReLU activation function:

In [ ]:
#build the model
model = Sequential()
# add 3 dense layers with 128, 64, 10 neurons with activation set to 'relu'
model.add(Dense(128, input_shape=(784,), activation="relu"))
model.add(Dense(64, activation="relu"))
model.add(Dense(10, activation="softmax"))
#compile the model
model.compile(optimizer='sgd', loss='categorical_crossentropy',
              metrics=['accuracy'])
#set 128 as mini batch size
model.fit(X_train, Y_train, batch_size=128, epochs=20, verbose=False)
#evaluate the model
score = model.evaluate(X_test, Y_test, verbose=0)
print('Test score:', score[0])
print('Test accuracy:', score[1])

Test score: 0.42280054092407227
Test accuracy: 0.8490999937057495


Using tanh activation function:

In [ ]:
#build the model
model = Sequential()
# add 3 dense layers with 128, 64, 10 neurons with activation set to tanh
model.add(Dense(128, input_shape=(784,), activation="tanh"))
model.add(Dense(64, activation="relu"))
model.add(Dense(10, activation="softmax"))
#compile the model
model.compile(optimizer='sgd', loss='categorical_crossentropy',
              metrics=['accuracy'])
#set 128 as mini batch size
model.fit(X_train, Y_train, batch_size=128, epochs=20, verbose=False)
#evaluate the model
score = model.evaluate(X_test, Y_test, verbose=0)
print('Test score:', score[0])
print('Test accuracy:', score[1])

Test score: 0.4173124432563782
Test accuracy: 0.8492000102996826


Using sigmoid activation function:

In [ ]:
#build the model
model = Sequential()
# add 3 dense layers with 128, 64, 10 neurons with activation set to sigmoid
model.add(Dense(128, input_shape=(784,), activation="sigmoid"))
model.add(Dense(64, activation="relu"))
model.add(Dense(10, activation="softmax"))
#compile the model
model.compile(optimizer='sgd', loss='categorical_crossentropy',
              metrics=['accuracy'])
#set 128 as mini batch size
model.fit(X_train, Y_train, batch_size=128, epochs=20, verbose=False)
#evaluate the model
score = model.evaluate(X_test, Y_test, verbose=0)

Test score: 0.499099999666214
Test accuracy: 0.8213000297546387


Varying the activation functions did produce slighly different results. The model that used the ReLU activation function had a very slightly higher accuracy score for the test set than the one that used the tanh activation function. These two models also had very similar scores for the training sets. The model with the sigmoid activation function had a lower accuracy score for the test set but a higher accuracy score for the training set. None of the three models appear to suffer from overfitting. The model that used the ReLU activation function did perform the best of the three, but only slightly.

Now, let's vary the number of neurons in each layer. I'll start with the model that has thus far performed the best and then vary the number of neurons.

In [ ]:
#build the model
model = Sequential()
# add 3 dense layers with 128, 64, 10 neurons with activation set to 'relu'
model.add(Dense(128, input_shape=(784,), activation="relu"))
model.add(Dense(64, activation="relu"))
model.add(Dense(10, activation="softmax"))
#compile the model
model.compile(optimizer='sgd', loss='categorical_crossentropy',
              metrics=['accuracy'])
#set 128 as mini batch size
model.fit(X_train, Y_train, batch_size=128, epochs=20, verbose=False)
#evaluate the model
score = model.evaluate(X_test, Y_test, verbose=0)
print('Test score:', score[0])
print('Test accuracy:', score[1])

Test score: 0.4214297831058502
Test accuracy: 0.8508999943733215


In [ ]:
#build the model
model = Sequential()
# add 3 dense layers with 100, 50, 10 neurons with activation set to 'relu'
model.add(Dense(100, input_shape=(784,), activation="relu"))
model.add(Dense(50, activation="relu"))
model.add(Dense(10, activation="softmax"))
#compile the model
model.compile(optimizer='sgd', loss='categorical_crossentropy',
              metrics=['accuracy'])
#set 128 as mini batch size
model.fit(X_train, Y_train, batch_size=128, epochs=20, verbose=False)
#evaluate the model
score = model.evaluate(X_test, Y_test, verbose=0)
print('Test score:', score[0])
print('Test accuracy:', score[1])

Test score: 0.4433915615081787
Test accuracy: 0.8460000157356262


In [ ]:
#build the model
model = Sequential()
# add 3 dense layers with 1000, 500, 10 neurons with activation set to 'relu'
model.add(Dense(1000, input_shape=(784,), activation="relu"))
model.add(Dense(500, activation="relu"))
model.add(Dense(10, activation="softmax"))
#compile the model
model.compile(optimizer='sgd', loss='categorical_crossentropy',
              metrics=['accuracy'])
#set 128 as mini batch size
model.fit(X_train, Y_train, batch_size=128, epochs=20, verbose=False)
#evaluate the model
score = model.evaluate(X_test, Y_test, verbose=0)
print('Test score:', score[0])
print('Test accuracy:', score[1])

Test score: 0.39236685633659363
Test accuracy: 0.861299991607666


In [ ]:
#build the model
model = Sequential()
# add 3 dense layers with 1500, 800, 10 neurons with activation set to 'relu'
model.add(Dense(1500, input_shape=(784,), activation="relu"))
model.add(Dense(800, activation="relu"))
model.add(Dense(10, activation="softmax"))
#compile the model
model.compile(optimizer='sgd', loss='categorical_crossentropy',
              metrics=['accuracy'])
#set 128 as mini batch size
model.fit(X_train, Y_train, batch_size=128, epochs=20, verbose=False)
#evaluate the model
score = model.evaluate(X_test, Y_test, verbose=0)
print('Test score:', score[0])
print('Test accuracy:', score[1])

Test score: 0.3887951076030731
Test accuracy: 0.8629000186920166


I was able to achieve a slightly higher accuracy when changing the number of neurons in the layers. Specifically, increasing the neurons in each layer to 1000, 500, and 10 resulted in a slightly higher accuracy score (86 %) but a slightly lower score for the training set. However, this model took longer to run and the 1% increase in accuracy is not worth the extra computational power needed.

Finally, I will tinker with the batch sizes to see how different models compare. I'll use a lower batch size number, a higher batch size, and the full sample.

In [ ]:
#build the model
model = Sequential()
# add 3 dense layers with 128, 64, 10 neurons with activation set to 'relu'
model.add(Dense(128, input_shape=(784,), activation="relu"))
model.add(Dense(64, activation="relu"))
model.add(Dense(10, activation="softmax"))
#compile the model
model.compile(optimizer='sgd', loss='categorical_crossentropy',
              metrics=['accuracy'])
#set 128 as mini batch size
model.fit(X_train, Y_train, batch_size=128, epochs=20, verbose=False)
#evaluate the model
score = model.evaluate(X_test, Y_test, verbose=0)
print('Test score:', score[0])
print('Test accuracy:', score[1])

Test score: 0.42241567373275757
Test accuracy: 0.8482000231742859


In [ ]:
#build the model
model = Sequential()
# add 3 dense layers with 128, 64, 10 neurons with activation set to 'relu'
model.add(Dense(128, input_shape=(784,), activation="relu"))
model.add(Dense(64, activation="relu"))
model.add(Dense(10, activation="softmax"))
#compile the model
model.compile(optimizer='sgd', loss='categorical_crossentropy',
              metrics=['accuracy'])
#set 300 as mini batch size
model.fit(X_train, Y_train, batch_size=300, epochs=20, verbose=False)
#evaluate the model
score = model.evaluate(X_test, Y_test, verbose=0)
print('Test score:', score[0])
print('Test accuracy:', score[1])

Test score: 0.47558721899986267
Test accuracy: 0.8331000208854675


In [ ]:
#build the model
model = Sequential()
# add 3 dense layers with 128, 64, 10 neurons with activation set to 'relu'
model.add(Dense(128, input_shape=(784,), activation="relu"))
model.add(Dense(64, activation="relu"))
model.add(Dense(10, activation="softmax"))
#compile the model
model.compile(optimizer='sgd', loss='categorical_crossentropy',
              metrics=['accuracy'])
#use full sample
model.fit(X_train, Y_train, batch_size=X_train.shape[0], epochs=20, verbose=False)
#evaluate the model
score = model.evaluate(X_test, Y_test, verbose=0)
print('Test score:', score[0])
print('Test accuracy:', score[1])

Test score: 2.0873217582702637
Test accuracy: 0.3075999915599823


Varying the batch size led to interesting results. Using the full sample produced a much lower accuracy for the test set (33.8%). Using a lower batch number (8) led to the highest accuracy for the test set thus far (89%) but also the lowest accuracy for the training set (suggesting that there might be some overfitting happening with that model).

Taking computational time and test/training accuracy rates into consideration, I would argue that the model with 3 dense layers, with 128, 64, and 10 neurons, with batch size set to 128, and with the ReLU activation function performed the best. This model does not seem to suffer from overfitting (based on the difference between the test/training set accuracy scores) and is fairly accuracy (85%). While a few models did have slightly higher accuracy scores, such models either had lower accuracy rates for the training sets (indicating potential overfitting issues) or took significantly longer to run.

In the checkpoints, we were able to achieve accuracy rates of 97% for the MNIST dataset using ANNs. In this challenge, I was only able to achieve accuracy rates of 80-88% even when varying the activation function, the number of layers, neurons, batch sizes, etc. This seems to indicate that it is more challenging to accurately model the Fashion MNIST data than the MNIST data. I could keep tuning the hyperparameters, but perhaps another deep learning approach could achieve even better accuracy rates than these ANN models were able to. 85% accuracy is fine but there is certainly room for improvement here.